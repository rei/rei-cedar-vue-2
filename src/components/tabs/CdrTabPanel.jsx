import clsx from 'clsx';
import modifier from '../../mixins/modifier';
import style from './styles/CdrTabPanel.scss';

export default {
  name: 'CdrTabPanel',
  mixins: [modifier],
  props: {
    /**
     * Required string value that shows up on tab header
     */
    name: {
      type: String,
      required: true,
    },
    disabled: {
      type: Boolean,
      default: false,
    },
    ariaLabelledby: {
      type: String,
      required: true,
    },
    id: {
      type: String,
      required: true,
    },
  },
  data() {
    return {
      active: false,
      hidden: true,
      offsetX: 0,
      animationDirection: null,
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tab-panel';
    },
    animationClass() {
      return this.animationDirection ? style[`cdr-tab-panel-${this.animationDirection}`] : null;
    },
  },
  methods: {
    setActive(state) {
      if (state) this.hidden = false;
      this.active = state;
      this.$emit('tab-change', state, this.id);
      this.$emit('tabChange', state, this.id);
    },
    setAnimationDirection(direction) {
      this.animationDirection = direction;
    },
    setOffsetX(x) {
      this.offsetX = x;
    },
    handleUpArrowNav() {
      this.$parent.setFocusToActiveTabHeader();
    },
    animationEnd(event) {
      if (event.animationName.split('-')[0] === 'exit') {
        this.hidden = true;
        this.animationDirection = null;
      }
    },
  },
  render() {
    return (
      <div
        aria-hidden={!this.active}
        aria-labelledby={this.ariaLabelledby}
        class={clsx(this.style[this.baseClass], this.modifierClass, this.animationClass)}
        hidden={this.hidden}
        id={this.id}
        ref="cdrTabPanelContainer"
        tabindex="0"
        role="tabpanel"
        vOn:keydown_up_prevent={this.handleUpArrowNav}
        vOn:animationend={this.animationEnd}
        key={this.name}
      >
        {this.$slots.default}
      </div>
    );
  },
};
