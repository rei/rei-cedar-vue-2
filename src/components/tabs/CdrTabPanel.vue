<template>
  <!-- disable lint errors on line length in template -->
  <!-- eslint-disable max-len -->
  <transition
    name="fly"
    v-on="animationHooks">
    <div
      v-show="active"
      :aria-hidden="!active"
      :class="[modifierClass]"
      tabindex="1"
      role="tabpanel"
      :key="name">
      <slot/>
    </div>
  </transition>
</template>

<script>
import modifier from 'mixinsdir/modifier';

export default {
  name: 'CdrTabPanel',
  mixins: [modifier],
  props: {
    name: String,
    id: String,
  },
  data() {
    return {
      active: false,
      offsetX: 0,
      tabId: this.id || this.name,
      animationDirection: 'default',
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tab-panel';
    },
    animationHooks() {
      return {
        beforeEnter: this.setEnterStart,
        afterEnter: this.setEnterEnd,
        beforeLeave: this.setLeaveStart,
        afterLeave: this.setLeaveEnd,
        ...this.$listeners,
      };
    },
  },
  methods: {
    setActive(state) {
      this.active = state;
      this.$emit('tabChange', state, this.tabId);
    },
    setAnimationDirection(direction) {
      this.animationDirection = direction;
    },
    setOffsetX(x) {
      this.offsetX = x;
    },
    setEnterStart(element) {
      const el = element;
      el.style.animationDirection = 'reverse';
      el.style.animationTimingFunction = 'cubic-bezier(0.4, 0, 0.68, .06)';
      el.classList.add(this.animationDirection);
    },
    setEnterEnd(element) {
      const el = element;
      el.style.animationDirection = '';
      el.classList.remove(this.animationDirection);
    },
    setLeaveStart(element) {
      const el = element;
      el.classList.add(this.animationDirection);
      el.style.animationTimingFunction = 'cubic-bezier(0.32, 0.94, 0.6, 1)';
    },
    setLeaveEnd(element) {
      const el = element;
      el.classList.remove(this.animationDirection);
    },
  },
};
</script>

<style>
  @import '../../css/settings/_index.pcss';
  @import './styles/CdrTabPanel.pcss';
</style>
