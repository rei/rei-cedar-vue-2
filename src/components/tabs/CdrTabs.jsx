import debounce from 'lodash-es/debounce';
import clsx from 'clsx';
import modifier from '../../mixins/modifier';
import size from '../../mixins/size';
import style from './styles/CdrTabs.scss';

export default {
  name: 'CdrTabs',
  mixins: [modifier, size],
  props: {
    height: {
      type: String,
      default: '240px',
    },
    activeTab: {
      type: Number,
      default: 0,
    },
  },
  data() {
    return {
      tabs: [],
      underlineOffsetX: 0,
      underlineWidth: 0,
      underlineScrollX: 0,
      activeTabIndex: 0,
      headerWidth: 0,
      headerOverflow: false,
      overflowLeft: false,
      overflowRight: false,
      animationInProgress: false,
      style,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tabs';
    },
    underlineStyle() {
      return {
        transform: `translateX(${this.underlineOffsetX}px)`,
        width: `${this.underlineWidth}px`,
      };
    },
  },
  mounted() {
    this.tabs = (this.$slots.default || [])
      .map(vnode => vnode.componentInstance)
      .filter(tab => tab); // get vue component children in the slot

    this.activeTabIndex = this.getNextTab(this.activeTab);

    if (this.tabs[this.activeTabIndex] && this.tabs[this.activeTabIndex].setActive) {
      this.tabs[this.activeTabIndex].setActive(true);
    }

    this.$nextTick(() => {
      this.headerWidth = this.getHeaderWidth();
      this.calculateOverflow();
      setTimeout(() => {
        this.updateUnderline();
      }, 100);
    });
    // Check for header overflow on window resize for gradient behavior.
    window.addEventListener('resize', debounce(() => {
      this.headerWidth = this.getHeaderWidth();
      this.calculateOverflow();
      this.updateUnderline();
    }, 500));
    // Check for header overflow on widow resize for gradient behavior.
    this.$refs.cdrTabsHeader.parentElement.addEventListener('scroll', debounce(() => {
      this.calculateOverflow();
      this.updateUnderline();
    }, 250));
  },
  methods: {
    getNextTab(startIndex) {
      for (let i = startIndex; i < this.tabs.length; i += 1) {
        if (!this.tabs[i].disabled) {
          return i;
        }
      }

      if (startIndex !== 0) {
        for (let k = 0; k < startIndex; k += 1) {
          if (!this.tabs[k].disabled) {
            return k;
          }
        }
      }

      return -1;
    },
    getPreviousTab(startIndex) {
      for (let i = startIndex; i > -1; i -= 1) {
        if (!this.tabs[i].disabled) {
          return i;
        }
      }

      if (startIndex !== this.tabs.length - 1) {
        for (let k = this.tabs.length - 1; k > startIndex; k -= 1) {
          if (!this.tabs[k].disabled) {
            return k;
          }
        }
      }

      return -1;
    },
    handleClick: debounce(function handleClickCallback(tabClicked) {
      const newIndex = this.tabs.findIndex(tab => tabClicked.name === tab.name);
      this.changeTab(newIndex);
    }, 500, { leading: true, trailing: false }),
    changeTab(newIndex) {
      const oldIndex = this.activeTabIndex;

      this.hideScrollBar();
      if (newIndex > oldIndex) {
        this.tabs[oldIndex].setAnimationDirection('exit-left');
        this.tabs[oldIndex].setActive(false);
        setTimeout(() => {
          this.tabs[newIndex].setActive(true);
          this.tabs[newIndex].setAnimationDirection('enter-right');
        }, 200);
      } else {
        this.tabs[oldIndex].setAnimationDirection('exit-right');
        this.tabs[oldIndex].setActive(false);
        setTimeout(() => {
          this.tabs[newIndex].setActive(true);
          this.tabs[newIndex].setAnimationDirection('enter-left');
        }, 200);
      }
      this.activeTabIndex = newIndex;
      this.updateUnderline();
      this.$refs.cdrTabsHeader.children[this.activeTabIndex].children[0].focus();
    },
    rightArrowNav: debounce(function handleRightArrow() {
      const nextTab = this.getNextTab(this.activeTabIndex + 1);
      if (nextTab !== -1) {
        this.changeTab(nextTab);
      }
    }, 300, { leading: true, trailing: false }),
    leftArrowNav: debounce(function handleLeftArrow() {
      const previousTab = this.getPreviousTab(this.activeTabIndex - 1);
      if (previousTab !== -1) {
        this.changeTab(previousTab);
      }
    }, 300, { leading: true, trailing: false }),
    calculateOverflow() {
      let containerWidth = 0;
      if (this.$refs.cdrTabsContainer) {
        containerWidth = this.$refs.cdrTabsContainer.offsetWidth;
      }
      this.headerOverflow = this.headerWidth > containerWidth;
      if (this.headerOverflow) {
        // Get Scroll Position
        const scrollX = this.$refs.cdrTabsHeader.parentElement.scrollLeft;
        this.overflowLeft = scrollX > 1;
        this.overflowRight = (scrollX + 1) < (this.headerWidth - containerWidth);
      } else {
        this.overflowLeft = false;
        this.overflowRight = false;
      }
    },
    updateUnderline() {
      const elements = Array.from(this.$refs.cdrTabsHeader.children);
      if (elements.length > 0) {
        const activeTab = elements[this.activeTabIndex];
        this.underlineOffsetX = activeTab.offsetLeft
          - this.$refs.cdrTabsHeader.parentElement.scrollLeft;
        this.underlineWidth = activeTab.firstChild.offsetWidth;
      }
    },
    handleDownArrowNav() {
      if (!this.animationInProgress) {
        this.$el.lastElementChild.children[this.activeTabIndex].focus();
      }
    },
    setFocusToActiveTabHeader() {
      this.$refs.cdrTabsHeader.children[this.activeTabIndex].children[0].focus();
    },
    getHeaderWidth() {
      let headerElements = [];
      if (this.$refs.cdrTabsHeader) {
        headerElements = Array.from(this.$refs.cdrTabsHeader.children);
      }
      let totalWidth = 0;
      headerElements.forEach((element) => {
        totalWidth += element.offsetWidth || 0;
      });
      return totalWidth;
    },
    hideScrollBar() {
      const styleRef = this.$refs.cdrTabsContainer.style;
      window.addEventListener('transitionend', () => {
        styleRef.setProperty('overflow-x', 'unset');
      }, { once: true });
      styleRef.setProperty('overflow-x', 'hidden');
    },
    getTabEl(tab) {
      return tab.disabled ? (
        <span
          class={clsx(
            this.style['cdr-tabs__header-item-label'],
            this.style['cdr-tabs__header-item-label--disabled'],
          )}
          aria-disabled="true"
          aria-selected="false"
        >
          {tab.name}
        </span>
      ) : (
        <a
          tabIndex={tab.active ? 0 : -1}
          vOn:click_prevent={e => this.handleClick(tab, e)}
          href={`#${tab.id}`}
          class={this.style['cdr-tabs__header-item-label']}
        >
          { tab.name }
        </a>
      );
    },
  },
  render() {
    return (
      <div
        class={clsx(this.style[this.baseClass], this.modifierClass, this.sizeClass)}
        ref="cdrTabsContainer"
        style={{ height: this.height }}
      >
        <div
          class={clsx(
            this.overflowLeft ? this.style['cdr-tabs__header-gradient-left'] : '',
            this.overflowRight ? this.style['cdr-tabs__header-gradient-right'] : '',
            this.style['cdr-tabs__gradient-container'],
          )}
          vOn:keyup_right={this.rightArrowNav}
          vOn:keyup_left={this.leftArrowNav}
          vOn:keydown_down_prevent={this.handleDownArrowNav}
        >
          <nav
            class={clsx(
              this.overflowLeft ? this.style['cdr-tabs__header-gradient-left'] : '',
              this.overflowRight ? this.style['cdr-tabs__header-gradient-right'] : '',
              this.style['cdr-tabs__header-container'],
            )}
          >
            <ol
              class={this.style['cdr-tabs__header']}
              role="tablist"
              ref="cdrTabsHeader"
            >
              {this.tabs.map(tab => (
                  <li
                    role="tab"
                    aria-selected={tab.active}
                    aria-disabled="false"
                    aria-controls={tab.id}
                    id={tab.ariaLabelledby}
                    key={tab.id}
                    class={clsx(
                      tab.active ? this.style['cdr-tabs__header-item-active'] : '',
                      this.style['cdr-tabs__header-item'],
                    )}
                  >
                    {this.getTabEl(tab)}
                  </li>
              ))}
            </ol>
          </nav>

          <div
            class={this.style['cdr-tabs__underline']}
            style={this.underlineStyle}
          />
        </div>
        <div
          class={this.style['cdr-tabs__content-container']}
          ref="slotWrapper"
        >
          {this.$slots.default}
        </div>
      </div>
    );
  },
};
