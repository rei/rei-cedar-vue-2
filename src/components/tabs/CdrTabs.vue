<template>
  <!-- disable lint errors on line length in template -->
  <!-- eslint-disable max-len -->
  <div
    :class="[modifierClass]"
    :style="{ height: height }"
    ref="cdrTabsContainer">
    <div
      :class="[ overflowLeft ? $style['cdr-tabs__header-gradient-left'] : '',
                overflowRight ? $style['cdr-tabs__header-gradient-right'] : '',
                $style['cdr-tabs__gradient-container']]"
      @keyup.right="handleArrowNav"
      @keyup.left="handleArrowNav">
      <nav
        :class="[ overflowLeft ? $style['cdr-tabs__header-gradient-left'] : '',
                  overflowRight ? $style['cdr-tabs__header-gradient-right'] : '',
                  $style['cdr-tabs__header-container']]">
        <ol
          :class="$style['cdr-tabs__header']"
          role="tablist"
          ref="cdrTabsHeader">
          <li
            v-for="tab in tabs"
            :key="tab.id"
            :class="[ tab.active ? $style['cdr-tabs__header-item-active'] : '', $style['cdr-tabs__header-item']]">
            <a
              role="tab"
              :tabindex="[ tab.active ? 1 : -1 ]"
              @click.prevent="handleClick(tab, $event)"
              :href="tab.name"
              :class="$style['cdr-tabs__header-item-label']">
              {{ tab.name }}
            </a>
          </li>
        </ol>
        <hr
          :class="$style['cdr-tabs__underline']"
          :style="underlineStyle">
      </nav>
    </div>
    <div :class="$style['cdr-tabs__content-container']">
      <slot/>
    </div>
  </div>
</template>

<script>
import modifier from 'mixinsdir/modifier';
import debounce from 'lodash/debounce';

export default {
  name: 'CdrTabs',
  mixins: [modifier],
  props: {
    height: {
      type: String,
      default: '240px',
    },
  },
  data() {
    return {
      tabs: [],
      underlineOffsetX: 0,
      underlineWidth: 0,
      widthInitialized: false,
      activeTabIndex: 0,
      headerOverflow: false,
      overflowLeft: false,
      overflowRight: false,
    };
  },
  computed: {
    baseClass() {
      return 'cdr-tabs';
    },
    underlineStyle() {
      return {
        marginLeft: `${this.underlineOffsetX}px`,
        width: `${this.underlineWidth}px`,
      };
    },
  },
  beforeMount() {
    this.tabs = this.$children;
  },
  mounted() {
    if (this.tabs[0] && this.tabs[0].setActive) this.tabs[0].setActive(true);
    // Check for header overflow on window resize for gradient behavior.
    window.addEventListener('resize', debounce(() => {
      this.headerWidth = this.getHeaderWidth();
      this.calculateOverflow();
    }, 250));
    // Check for header overflow on widow resize for gradient behavior.
    this.$refs.cdrTabsHeader.parentElement.addEventListener('scroll', debounce(() => {
      this.calculateOverflow();
    }, 250));
    this.headerWidth = this.getHeaderWidth();
  },
  updated() {
    this.initializeOffsets();
  },
  methods: {
    handleClick(tabClicked, event) {
      const newSelectedTab = this.tabs.find(tab => tabClicked.name === tab.name);
      this.tabs.forEach((tab, index) => {
        if (newSelectedTab.name === tab.name) {
          if (this.activeTabIndex < index) {
            tab.setAnimationDirection('flyRight');
            this.tabs[this.activeTabIndex].setAnimationDirection('flyLeft');
          } else {
            tab.setAnimationDirection('flyLeft');
            this.tabs[this.activeTabIndex].setAnimationDirection('flyRight');
          }
          this.activeTabIndex = index;
          this.$nextTick(tab.setActive(true));
        } else {
          this.$nextTick(tab.setActive(false));
        }
      });
      this.underlineOffsetX =
        event.currentTarget.offsetLeft
        - event.currentTarget.parentElement.parentElement.offsetLeft;
      this.underlineWidth = event.currentTarget.offsetWidth;
    },
    initializeOffsets() {
      if (!this.widthInitialized) {
        const elements = Array.from(this.$refs.cdrTabsHeader.children);
        this.underlineWidth = elements[0].children[0].offsetWidth;
        this.widthInitialized = true;
        // Set focus to default Tab header
        this.$nextTick(() => {
          this.$refs.cdrTabsHeader.children[this.activeTabIndex].children[0].focus();
        });
      }
    },
    calculateOverflow() {
      const containerWidth = this.$refs.cdrTabsContainer.offsetWidth;
      this.headerOverflow = this.headerWidth > containerWidth;
      if (this.headerOverflow) {
        // Get Scroll Position
        const scrollX = this.$refs.cdrTabsHeader.parentElement.scrollLeft;
        this.overflowLeft = scrollX > 1;
        this.overflowRight = (scrollX + 1) < (this.headerWidth - containerWidth);
      } else {
        this.overflowLeft = false;
        this.overflowRight = false;
      }
    },
    handleArrowNav(event) {
      if (event.which === 39) {
        // navigate right
        if (this.activeTabIndex < (this.tabs.length - 1)) {
          this.tabs[this.activeTabIndex].setAnimationDirection('flyLeft');
          this.tabs[this.activeTabIndex + 1].setAnimationDirection('flyRight');
          this.$nextTick(this.tabs[this.activeTabIndex].setActive(false));
          this.activeTabIndex += 1;
          this.$nextTick(this.tabs[this.activeTabIndex].setActive(true));
        }
      } else if (event.which === 37) {
        // navigate left
        if (this.activeTabIndex > 0) {
          this.tabs[this.activeTabIndex].setAnimationDirection('flyRight');
          this.tabs[this.activeTabIndex - 1].setAnimationDirection('flyLeft');
          this.$nextTick(this.tabs[this.activeTabIndex].setActive(false));
          this.activeTabIndex -= 1;
          this.$nextTick(this.tabs[this.activeTabIndex].setActive(true));
        }
      }
      if (this.$refs.cdrTabsHeader.children[this.activeTabIndex] &&
        (event.which === 37 || event.which === 39)) {
        this.underlineOffsetX =
        this.$refs.cdrTabsHeader.children[this.activeTabIndex].offsetLeft
          - this.$refs.cdrTabsHeader.offsetLeft;
        this.underlineWidth =
          this.$refs.cdrTabsHeader.children[this.activeTabIndex].children[0].offsetWidth;
        this.$refs.cdrTabsHeader.children[this.activeTabIndex].children[0].focus();
      }
    },
    getHeaderWidth() {
      const headerElements = Array.from(this.$refs.cdrTabsHeader.children);
      let totalWidth = 0;
      headerElements.forEach((element) => {
        totalWidth += element.offsetWidth || 0;
      });
      return totalWidth;
    },
  },
};
</script>

<style module>
  @import '../../css/settings/_index.pcss';
  @import './styles/CdrTabs.pcss';
</style>
