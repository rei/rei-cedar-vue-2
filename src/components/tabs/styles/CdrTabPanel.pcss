/* ==========================================================================
  # CdrTabPanel

  TOC:

    :Base - CdrTabPanel

========================================================================== */

.cdr-tab-panel {
  position: absolute;
  width: 100%;
  height: 100%;
}

@keyframes flyLeft {
  from {
    transform: translateX(0);
  }

  to {
    transform: translateX(-100vw);
  }
}

@keyframes flyRight {
  from {
    transform: translateX(0);
  }

  to {
    transform: translateX(100vw);
  }
}

.flyLeft {
  animation-duration: 0.5s;

  /* animation-timing-function: cubic-bezier(0.32, 0.94, 0.6, 1); */
  animation-name: flyLeft;
}

.flyRight {
  animation-duration: 0.5s;

  /* animation-timing-function: cubic-bezier(0.32, 0.94, 0.6, 1); */
  animation-name: flyRight;
}
