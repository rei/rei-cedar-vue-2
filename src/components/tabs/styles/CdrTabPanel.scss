@import '../../../css/settings/index.scss';

.cdr-tab-panel {
  position: absolute;
  width: 100%;
  height: 100%;
  padding-top: $cdr-space-eighth-x;
}

@keyframes :global(exit-left) {
  from {
    transform: translateX(0);
    opacity: 1;
  }
  
  to {
    transform: translateX(-10px);
    opacity: 0;
  }
}

@keyframes :global(exit-right) {
  from {
    transform: translateX(0);
    opacity: 1;
  }

  to {
    transform: translateX(10px);
    opacity: 0;
  }
}

@keyframes :global(enter-left) {
  from {
    transform: translateX(-10px);
    opacity: 0;
  }

  to {
    transform: translateX(0);
    opacity: 1;
  }
}

@keyframes :global(enter-right) {
  from {
    transform: translateX(10px);
    opacity: 0;
  }

  to {
    transform: translateX(0);
    opacity: 1;
  }
}

@mixin cdr-tab-panel-animation {
  animation-duration: $cdr-duration-2-x;
  animation-timing-function: $cdr-timing-function-ease-out;
}

.cdr-tab-panel-enter-left {
  :global {
    animation-name: enter-left;
  }
  @include cdr-tab-panel-animation;
}

.cdr-tab-panel-enter-right {
  :global {
    animation-name: enter-right;
  }
  @include cdr-tab-panel-animation;
}

.cdr-tab-panel-exit-left {
  :global {
    animation-name: exit-left;
  }
  @include cdr-tab-panel-animation;
}

.cdr-tab-panel-exit-right {
  :global {
    animation-name: exit-right;
  }
  @include cdr-tab-panel-animation;
}

