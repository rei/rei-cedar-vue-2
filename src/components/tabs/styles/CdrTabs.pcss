/* ==========================================================================
  # CdrTabs

  All values should map to variables in
  vars/CdrTabs.vars.pcss
  in order to allow for theming

  TOC:

    :Base - CdrTabs
    :Elements
      :header
      :header-item
      :header-item-button
      :underline
      :content-container
    :Modifiers
      :compact
      :no-border
      :full-width
========================================================================== */

.cdr-tabs {
  /* overflow-x: auto; */
  display: flex;
  flex-flow: column;
  height: 500px;

  &__header-container {
    overflow-x: auto;
    border-bottom: 1px solid $grey-matter;
    -webkit-overflow-scrolling: touch;
    flex: 0 1 auto;
    -ms-overflow-style: -ms-autohiding-scrollbar;
  }

  &__gradient-container {
    position: relative;

    &::before {
      transition: opacity 150ms ease;
      -webkit-transition: opacity 150ms ease;
      content: '';
      position: absolute;
      z-index: 100;
      top: 0;
      bottom: 0;
      left: 0;
      width: 36px;
      background:
        linear-gradient(
          to left,
          rgba(255, 255, 255, 0),
          $clean-slate
        );
      pointer-events: none;
      opacity: 0;
    }

    &::after {
      transition: opacity 150ms ease;
      -webkit-transition: opacity 150ms ease;
      content: '';
      position: absolute;
      top: 0;
      bottom: 0;
      right: 0;
      width: 36px;
      background:
        linear-gradient(
          to right,
          rgba(255, 255, 255, 0),
          $clean-slate
        );
      pointer-events: none;
      opacity: 0;
    }
  }

  &__header {
    display: inline-flex;
    width: 100%;
    margin: 0;

    /* align-content: stretch; */
    padding: 0;
    position: relative;

    &-gradient-left::before {
      opacity: 1;
    }

    &-gradient-right::after {
      opacity: 1;
    }
  }

  &__header-item {
    display: inline-flex;
    padding: 0;

    &[aria-selected] > .cdr-tabs__header-item-label {
      color: $easily-excited;
    }
  }

  &__header-item-label {
    @include redwood-display-30;

    text-decoration: none;
    margin: 0 $space-1-x 0 0;
    padding: $inset-half-x-stretch;
    color: $heart-of-darkness;
    white-space: nowrap;
    outline-offset: -3px;

    &:active,
    &:hover,
    &:focus {
      text-decoration: none;
      color: $easily-excited;
    }
  }

  &__underline {
    position: absolute;
    width: 30px;
    margin: 0;
    height: $space-quarter-x;
    box-sizing: border-box;
    border: none;

    /* border: none; */
    background: $easily-excited;
    transition: 0.6s cubic-bezier(0.32, 0.94, 0.6, 1);
  }

  &__content-container {
    position: relative;
    flex: 1 1 auto;
    overflow: auto;
  }

  .fade-enter-active,
  .fade-leave-active {
    opacity: 0.5s;
  }

  .fade-enter,
  .fade-leave-to {
    opacity: 0;
  }

  /* Compact
    ========== */
  &--compact {
    .cdr-tabs__header-item-label {
      @include redwood-display-20;

      margin: 0 $space-half-x 0 0;
      padding: $inset-three-quarter-x-squish;
    }
  }

  /* No Border
    ========== */
  &--no-border {
    .cdr-tabs__header,
    .cdr-tabs__header-container {
      border-bottom: 1px solid transparent;
    }
  }

  /* Full Width
    ========== */
  &--full-width {
    .cdr-tabs__header {
      justify-content: space-between;
    }
  }
}
