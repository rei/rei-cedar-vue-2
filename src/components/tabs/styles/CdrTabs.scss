@import '../../../css/settings/index.scss';
@import url('@rei/cedar/dist/style/cdr-tab-panel.css');

@mixin cdr-tabs-base-header-item-label() {
  font-family: $cdr-font-family-sans;
  font-style: normal;
  font-weight: 500;
  letter-spacing: -.08px;
  font-size: 1.6rem;
  line-height: 2.2rem;
}

@mixin cdr-tabs-small-header-item-label() {
  font-family: $cdr-font-family-sans;
  font-style: normal;
  font-weight: 500;
  letter-spacing: -.08px;
  font-size: 1.4rem;
  line-height: 1.8rem;
}

.cdr-tabs {
  /* overflow-x: auto; */
  display: flex;
  flex-flow: column;
  height: 500px;

  // https://davidwalsh.name/osx-overflow
  ::-webkit-scrollbar {
    -webkit-appearance: none;
    width: 7px;
  }
  ::-webkit-scrollbar-thumb {
    border-radius: 4px;
    background-color: rgba(0, 0, 0, .5);
    -webkit-box-shadow: 0 0 1px rgba(255, 255, 255, .5);
  }

  &__header-container {
    overflow-x: auto;
    border-bottom: 1px solid $cdr-color-border-tab-keyline-rest;
    -webkit-overflow-scrolling: touch;
    -ms-overflow-style: -ms-autohiding-scrollbar;
    &::-webkit-scrollbar { width: 0 !important; height: 0 !important; }
  }

  &__gradient-container {
    position: relative;

    &::before {
      transition: opacity 150ms ease;
      -webkit-transition: opacity 150ms ease;
      content: '';
      position: absolute;
      z-index: 100;
      top: 0;
      bottom: 0;
      left: 0;
      width: 36px;
      background:
        linear-gradient(
          to left,
          rgba($cdr-color-background-lightest, 0),
          $cdr-color-background-lightest
        );
      pointer-events: none;
      opacity: 0;
    }

    &::after {
      transition: opacity 150ms ease;
      -webkit-transition: opacity 150ms ease;
      content: '';
      position: absolute;
      top: 0;
      bottom: 0;
      right: 0;
      width: 36px;
      background:
        linear-gradient(
          to right,
          rgba($cdr-color-background-lightest, 0),
          $cdr-color-background-lightest
        );
      pointer-events: none;
      opacity: 0;
    }
  }

  &__header {
    display: flex;
    width: 100%;
    margin: 0;

    padding: 0;
    position: relative;

    &-gradient-left::before {
      opacity: 1;
    }

    &-gradient-right::after {
      opacity: 1;
    }
  }

  &__header-item {
    display: block;
    padding: 0;

    & + & {
      margin-left: $cdr-space-one-x;
    }

    & [aria-selected="true"].cdr-tabs__header-item-label {
      color: $cdr-color-text-tab-active;
      font-weight: 500;
    }
  }

  &__header-item-label {
    @include cdr-tabs-base-header-item-label;

    display: block;
    text-decoration: none;
    padding: $cdr-space-inset-half-x-stretch;
    color: $cdr-color-text-tab-rest;
    white-space: nowrap;
    outline-offset: -3px;

    &:active,
    &:hover,
    &:focus {
      text-decoration: none;
      color: $cdr-color-text-tab-hover;
    }

    &--disabled {
      color: $cdr-color-text-tab-disabled;

      &:active,
      &:hover,
      &:focus {
        color: $cdr-color-text-tab-disabled;
        cursor: not-allowed;
      }
    }
  }

  &__underline {
    z-index: 1;
    position: absolute;
    width: 30px;
    margin: 0;
    height: $cdr-space-quarter-x;
    margin-top: -$cdr-space-quarter-x;
    box-sizing: border-box;
    border: none;

    /* border: none; */
    background-color: $cdr-color-text-link-lightmode;
    transition: $cdr-duration-4-x $cdr-timing-function-ease-out;
  }

  &__content-container {
    position: relative;
    flex: 1 1 auto;
    overflow-x: hidden;
    overflow-y: auto;
  }

  .fade-enter-active,
  .fade-leave-active {
    opacity: 0.5s;
  }

  .fade-enter,
  .fade-leave-to {
    opacity: 0;
  }

  /* Compact
    ========== */
  &--compact, &--small {
    .cdr-tabs__header-item-label {
      @include cdr-tabs-small-header-item-label;

      padding: $cdr-space-inset-three-quarter-x-squish;

    }
    .cdr-tabs__header-item {
      & + & {
        margin-left: $cdr-space-half-x;
      }
    }
  }

  /* No Border
    ========== */
  &--no-border {
    .cdr-tabs__header,
    .cdr-tabs__header-container {
      border-bottom: 1px solid transparent;
    }
  }

  /* Full Width
    ========== */
  &--full-width {
    .cdr-tabs__header-item {
      flex-grow: 1;
    }
  }

  /* Centered
    ========== */
  &--centered {
    .cdr-tabs__header-item:first-of-type {
      margin-left: auto;
    }
    .cdr-tabs__header-item:last-of-type {
      margin-right: auto;
    }
  }
}
