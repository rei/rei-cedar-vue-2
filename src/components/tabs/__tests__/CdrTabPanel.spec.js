import { shallowMount, mount } from '@vue/test-utils';
import CdrTabPanel from 'componentdir/tabs/CdrTabPanel';
import CdrTabs from 'componentdir/tabs/CdrTabs';
import Vue from 'vue';


describe('CdrTabPanel', () => {
  it('renders tab', () => {
    const wrapper = mount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
      },
    });
    expect(wrapper.element).toMatchSnapshot();
  });

  it('is not active by default', () => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
      },
    });
    expect(wrapper.vm.active).toBe(false);
  });

  it('is active when set', (done) => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
      },
    });
    wrapper.vm.setActive(true);
    wrapper.vm.$nextTick(() => {
      expect(wrapper.vm.active).toBe(true);
      done();
    });
  });

  it('set animation direction functions correctly', (done) => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
      },
    });
    wrapper.vm.setActive(true);

    wrapper.vm.$nextTick(() => {
      wrapper.vm.setAnimationDirection('flyLeft');
      expect(wrapper.vm.animationDirection).toBe('flyLeft');
      done();
    });
  });

  it('set animation direction functions correctly', (done) => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
      },
    });
    wrapper.vm.setActive(true);

    wrapper.vm.$nextTick(() => {
      wrapper.vm.setOffsetX(1234);
      expect(wrapper.vm.offsetX).toBe(1234);
      done();
    });
  });

  it('updates state after animationend', () => {
    const wrapper = shallowMount(CdrTabPanel, {
      propsData: {
        name: 'test',
        id: 'tab1',
      },
    });

    wrapper.setData({ active: true, hidden: false, animationDirection: 'exit-left' });
    wrapper.vm.$nextTick(() => {
      wrapper.trigger('animationend', {
        animationName: 'exit-left'
      });
      expect(wrapper.vm.hidden).toBe(true);
      expect(wrapper.vm.animationDirection).toBe(null);
    });

  });

  it('handleUpArrowNav', () => {
    const spyUpdateUnderline = jest.fn();
    const spySetFocusToActiveTabHeader = jest.fn();
    const wrapper = mount(CdrTabs, {
      stubs: {
        'cdr-tab-panel': CdrTabPanel,
      },
      slots: {
        default: ['<cdr-tab-panel name="tab1" id="tab1" />', '<cdr-tab-panel name="tab2" id="tab2" />']
      },
      methods: {
        updateUnderline: spyUpdateUnderline,
        setFocusToActiveTabHeader: spySetFocusToActiveTabHeader,
      },
      attachToDocument: true,
    });

    Vue.nextTick(() => {
      wrapper.find(CdrTabPanel).trigger('keydown.up');
      expect(spySetFocusToActiveTabHeader).toHaveBeenCalled();
    });
  })
});