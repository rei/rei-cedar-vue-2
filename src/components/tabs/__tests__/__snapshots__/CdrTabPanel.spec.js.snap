// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CdrTabPanel renders tab 1`] = `
<div
  aria-hidden="true"
  class="cdr-tab-panel"
  name="fly"
  role="tabpanel"
  style="display: none;"
  tabindex="0"
/>
`;
